### PR TITLE
GH-661 Add support for passing global properties to applications

### PR DESCRIPTION
**Pull request by mbogoevici**
Resolves spring-cloud/spring-cloud-dataflow#661

### CODE REVIEW

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/StreamDeploymentController.java [hunk 1]
```diff
@@ -103,19 +104,27 @@
...
+	private final CommonApplicationProperties appsProperties;
+
+	/**
 	 * Create a {@code StreamDeploymentController} that delegates
 	 * <ul>
 	 * <li>CRUD operations to the provided {@link StreamDefinitionRepository}</li>
 	 * <li>app retrieval to the provided {@link AppRegistry}</li>
 	 * <li>deployment operations to the provided {@link AppDeployer}</li>
 	 * </ul>
+<<<<<<< HEAD
```
Comment (ilayaperumalg) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/StreamDeploymentController.java]: This needs to go ;)

Comment (mbogoevici) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/StreamDeploymentController.java]: Snap :(

File: spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/configuration/TestDependencies.java [hunk 2]
```diff
@@ -65,25 +68,30 @@
 @EnableHypermediaSupport(type = HAL)
 @Import(CompletionConfiguration.class)
 @EnableWebMvc
+@EnableConfigurationProperties(CommonApplicationProperties.class)
 public class TestDependencies extends WebMvcConfigurationSupport {
 
 	private final MavenProperties mavenProperties = new MavenProperties();
 
+	@Autowired
```
Comment (ilayaperumalg) [inline spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/configuration/TestDependencies.java]: Not sure what is the reasoning behind `autowiring` here. May be we just want to initialize the `CommonApplicationProperties` without having to do `@EnableConfigurationProperties`?

Comment (mbogoevici) [inline spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/configuration/TestDependencies.java]: Actually. It's the initialization that doesn't make sense. The test modifies the properties bean.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/apps/CommonApplicationProperties.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.springframework.boot.context.properties.ConfigurationProperties;
+
+/**
+ * Common properties for applications deployed via Spring Cloud Data Flow.
+ *
+ * @author Marius Bogoevici
+ */
+@ConfigurationProperties("spring.cloud.dataflow.applicationProperties")
```
Comment (ilayaperumalg) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/apps/CommonApplicationProperties.java]: We need to document this and created one for this: https://github.com/spring-cloud/spring-cloud-dataflow/issues/690

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/StreamDeploymentController.java (hunk 1 above)
Comment (ilayaperumalg) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/StreamDeploymentController.java]: will fix while merging.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/StreamDeploymentController.java [hunk 4]
```diff
@@ -300,24 +310,20 @@ private StreamAppDefinition qualifyParameters(StreamAppDefinition original, Reso
 		String wildCardConsumerPropertyPrefix = "app.*.consumer.";
 		String wildCardPrefix = "app.*.";
 		// first check for wild card prefix
```
Comment (ilayaperumalg) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/StreamDeploymentController.java]: This comment needs to be placed next line.

File: spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/controller/StreamControllerTests.java [hunk 5]
```diff
@@ -572,6 +580,40 @@ public void testDeployWithWildcardProperties() throws Exception {
 	}
 
 	@Test
+	public void testDeployWithCommonApplicationProperties() throws Exception {
+		repository.save(new StreamDefinition("myStream", "time | log"));
+		assertThat(appsProperties.getStream().values(), empty());
+		appsProperties.getStream().put("spring.cloud.stream.fake.binder.host","fakeHost");
```
Comment (mbogoevici) [inline spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/controller/StreamControllerTests.java]: On second thought, we shouldn't do this, as there may be multiple tests modifying the state. Ideally, we should create a new deployer with common properties initialized.

Comment (mbogoevici) [inline spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/controller/StreamControllerTests.java]: On the other hand, the tests already change the state of the repository, registering different stream definitions, so they couldn't normally run in parallel anyway. So probably this could be left like this.

File: spring-cloud-dataflow-docs/src/main/asciidoc/streams.adoc [hunk 6]
```diff
@@ -233,6 +233,21 @@ configuration.classes=org.springframework.cloud.stream.app.file.sink.FileSinkPro
 configuration-properties.names=server.port
 ```
 
+=== Global application properties (for streams)
```
Comment (mbogoevici) [inline spring-cloud-dataflow-docs/src/main/asciidoc/streams.adoc]: could ditch the paranthesis

File: spring-cloud-dataflow-docs/src/main/asciidoc/streams.adoc [hunk 7]
```diff
@@ -233,6 +233,21 @@ configuration.classes=org.springframework.cloud.stream.app.file.sink.FileSinkPro
...
+
+In addition to configuration via DSL, Spring Cloud Data Flow provides a mechanism for setting specific properties to all the streaming applications that are launched by it.
+This can be done by adding properties prefixed with `spring.cloud.dataflow.applicationProperties.stream` when starting the server.
+When doing so, the server will pass all the properties, without the prefix, to the instances it launches.
+
+For example, all the launched applications can be reconfigured to use a different Kafka broker by the default by launching the configuration server with the following options:
```
Comment (ilayaperumalg) [inline spring-cloud-dataflow-docs/src/main/asciidoc/streams.adoc]: @mbogoevici Given these common application properties are for the dataflow server and launching the `config` server with the new properties would cause the refresh on the dataflow server configuration only. Is that right? In that case, how can we expect the deployed apps to be refreshed with these properties. what am I missing?

File: spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/controller/StreamControllerTests.java (hunk 5 above)
Comment (markfisher) [inline spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/controller/StreamControllerTests.java]: Maybe just reinitialize `appsProperties` in a `@Before` method?

File: spring-cloud-dataflow-docs/src/main/asciidoc/streams.adoc (hunk 7 above)
Comment (mbogoevici) [inline spring-cloud-dataflow-docs/src/main/asciidoc/streams.adoc]: I don't think I mentioned refresh anywhere. We can simply replace `reconfigured` with `configured` and it'll make more sense, I guess?

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/apps/CommonApplicationProperties.java [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+ * Common properties for applications deployed via Spring Cloud Data Flow.
+ *
+ * @author Marius Bogoevici
+ */
+@ConfigurationProperties("")
+public class CommonApplicationProperties {
+
+	private Map<String,String> stream = new ConcurrentHashMap<>();
+
+	public Map<String, String> getStream() {
```
Comment (markfisher) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/apps/CommonApplicationProperties.java]: I presume the plan is to add `task` as well?

File: spring-cloud-dataflow-docs/src/main/asciidoc/streams.adoc (hunk 7 above)
Comment (mbogoevici) [inline spring-cloud-dataflow-docs/src/main/asciidoc/streams.adoc]: e.g. "For example, all the launched applications can be configured to use a specific Kafka broker by launching the configuration server with the following options:"

File: spring-cloud-dataflow-docs/src/main/asciidoc/streams.adoc [hunk 9]
```diff
@@ -233,6 +233,21 @@ configuration.classes=org.springframework.cloud.stream.app.file.sink.FileSinkPro
...
+
+```
+--spring.cloud.dataflow.applicationProperties.stream.spring.cloud.stream.kafka.binder.brokers=192.168.1.100:9092
+--spring.cloud.dataflow.applicationProperties.stream.spring.cloud.stream.kafka.binder.zkNodes=192.168.1.100:2181
+```.
+
+This will cause the properties `stream.spring.cloud.stream.kafka.binder.brokers` and `spring.cloud.stream.kafka.binder.zkNodes` to be passed to all the launched applications.
```
Comment (markfisher) [inline spring-cloud-dataflow-docs/src/main/asciidoc/streams.adoc]: Should mention here that IF the same property is provided explicitly with the `app.<name>.` prefix, that value would take precedence. And, on that note, would we possibly want a boolean flag to control that precedence? (thinking of the case where config-server might be used, and overriding explicit property values might be desirable).

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/apps/CommonApplicationProperties.java (hunk 8 above)
Comment (mbogoevici) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/apps/CommonApplicationProperties.java]: Eventually, but I'm not overly familiar with how task properties are set. I suppose that there are properties that are common to both, too.

Comment (markfisher) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/apps/CommonApplicationProperties.java]: we should make sure we're ok with the prefix namespaces that we'll support (especially if there is a group of properties that would apply to either stream or task apps) cc @mminella

File: spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/controller/StreamControllerTests.java (hunk 5 above)
Comment (mbogoevici) [inline spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/controller/StreamControllerTests.java]: Not a concern if we leave `@EnableConfigurationProperties` in, since the context is marked dirty after each method, I guess?

File: spring-cloud-dataflow-docs/src/main/asciidoc/streams.adoc (hunk 7 above)
Comment (markfisher) [inline spring-cloud-dataflow-docs/src/main/asciidoc/streams.adoc]: this does raise an interesting point... the apps themselves could be reading properties from `config-server` (not just indirectly via dataflow deploying the apps), including support for refreshable properties if implemented as such by the stream app developer

File: spring-cloud-dataflow-docs/src/main/asciidoc/streams.adoc (hunk 9 above)
Comment (mbogoevici) [inline spring-cloud-dataflow-docs/src/main/asciidoc/streams.adoc]: Config server would aways take precedence, IIRC.

File: spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/controller/StreamControllerTests.java (hunk 5 above)
Comment (markfisher) [inline spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/controller/StreamControllerTests.java]: right... it would be autowired each time due to `@DirtiesContext` being present on this test class

File: spring-cloud-dataflow-docs/src/main/asciidoc/streams.adoc (hunk 9 above)
Comment (markfisher) [inline spring-cloud-dataflow-docs/src/main/asciidoc/streams.adoc]: but here the "common" properties are being set first, and then the `app.<name>.` properties are set next... thus overriding, right?: https://github.com/spring-cloud/spring-cloud-dataflow/pull/688/files#diff-e6925c2ec0e82e9d426e01d83d3051b1R313

File: spring-cloud-dataflow-docs/src/main/asciidoc/streams.adoc (hunk 7 above)
Comment (mbogoevici) [inline spring-cloud-dataflow-docs/src/main/asciidoc/streams.adoc]: Sure, but that's a different concern (and we'd need to add spring-cloud-bus to starter apps as well). In this case it was just infortunate that the initial phrasing could be somehow misconstrued.

File: spring-cloud-dataflow-docs/src/main/asciidoc/streams.adoc (hunk 9 above)
Comment (mbogoevici) [inline spring-cloud-dataflow-docs/src/main/asciidoc/streams.adoc]: Yes, it helps to explain that the "common" properties are lowest in precedence among the ones set explicitly (which is to be expected) - I think that's intuitive enough, wdyt? - e.g. "common" -> "app.*" -> "app.name".

What I meant was that config server is orthogonal and typically takes precedence due to the priority of its property source.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/apps/CommonApplicationProperties.java (hunk 8 above)
Comment (mminella) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/apps/CommonApplicationProperties.java]: I'm not sure I see how this applies on the task side.  Since a task is a singular thing from the perspective of SCDF, global properties would be the same as regular properties.  Tasks do have the ability to launch other tasks, but that's a function of the task itself and not of SCDF.

File: spring-cloud-dataflow-docs/src/main/asciidoc/streams.adoc (hunk 9 above)
Comment (markfisher) [inline spring-cloud-dataflow-docs/src/main/asciidoc/streams.adoc]: but do we foresee any case where it might be desirable for the common properties to take precedence even over those provided via `stream deploy` ?

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/apps/CommonApplicationProperties.java (hunk 8 above)
Comment (markfisher) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/apps/CommonApplicationProperties.java]: @mminella I believe this could still be relevant for tasks just as a convenience... i.e. not having to provide common properties every time a task is launched

File: spring-cloud-dataflow-docs/src/main/asciidoc/streams.adoc (hunk 9 above)
Comment (mbogoevici) [inline spring-cloud-dataflow-docs/src/main/asciidoc/streams.adoc]: Honestly, I don't believe so - and certainly can't see a case where that would be controlled by yet another global setting. I believe that the rule that "the more general the property, the lower precedence it has would work just fine" - it's a simple and quite intuitive rule.

Comment (markfisher) [inline spring-cloud-dataflow-docs/src/main/asciidoc/streams.adoc]: ok

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/apps/CommonApplicationProperties.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.springframework.boot.context.properties.ConfigurationProperties;
+
+/**
+ * Common properties for applications deployed via Spring Cloud Data Flow.
+ *
+ * @author Marius Bogoevici
+ */
+@ConfigurationProperties("")
```
Comment (mbogoevici) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/apps/CommonApplicationProperties.java]: Ugh. This got somehow deleted.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/apps/CommonApplicationProperties.java (hunk 3 above)
Comment (mbogoevici) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/apps/CommonApplicationProperties.java]: Documentation pushed as part of the PR.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/StreamDeploymentController.java [hunk 11]
```diff
@@ -126,6 +134,7 @@ public StreamDeploymentController(StreamDefinitionRepository repository, Deploym
 		this.registry = registry;
 		this.deployer = deployer;
 		this.metadataResolver = metadataResolver;
+		this.appsProperties = commonProperties;
```
Comment (markfisher) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/StreamDeploymentController.java]: we should assert this is not null, since otherwise we could hit an NPE on line 309

**Comment by markfisher** [comment]
@mbogoevici can you please rebase this PR? (and either you or I, while merging, can add that one `assertNotNull`)... there are conflicts due to recent doc updates

**Comment by mbogoevici** [comment]
Squashed, rebased on top of master. Addressed comments.

**Comment by markfisher** [comment]
LGTM; merged
